File: src/main/java/fortress/tfol/operations/DeBruijnConverter.java
package fortress.tfol.operations;

import fortress.tfol.*;

import java.util.LinkedList;
import java.util.stream.Collectors;
import java.util.List;
import java.util.ArrayList;

public class DeBruijnConverter {
    private int counter;
    private LinkedList<Mapping> mappingStack;
    private DeBruijnVisitor visitor;
    
    private class Mapping {
        private int index;
        private Var variable;
        
        public Mapping(int index, Var variable) {
            this.index = index;
            this.variable = variable;
        }
    }
    
    public DeBruijnConverter() {
        this.counter = 0;
        this.mappingStack = new LinkedList<>();
        this.visitor = new DeBruijnVisitor();
    }
    
    public Term convert(Term term) {
        this.counter = 0;
        this.mappingStack = new LinkedList<>();
        return visitor.visit(term);
    }
    
    private class DeBruijnVisitor implements TermVisitor<Term> {
        
        @Override
        public Term visitTop(Top top) {
            return top;
        }
        
        @Override
        public Term visitBottom(Bottom bottom) {
            return bottom;
        }
        
        @Override
        public Term visitVar(Var var) {
            for(Mapping m : mappingStack) {
                if(var.equals(m.variable)) {
                    return Term.mkVar("_" + Integer.toString(m.index));
                }
            }
            return var;
        }
        
        @Override
        public Term visitNot(Not not) {
            return not.mapBody(this::visit);
        }
        
        @Override
        public Term visitAndList(AndList and) {
            return and.mapArguments(this::visit);
        }
        
        @Override
        public Term visitOrList(OrList or) {
            return or.mapArguments(this::visit);
        }
        
        @Override
        public Term visitDistinct(Distinct dist) {
            return dist.mapArguments(this::visit);
        }
        
        @Override
        public Term visitImplication(Implication imp) {
            return imp.mapArguments(this::visit);
        }
        
        @Override
        public Term visitIff(Iff iff) {
            return iff.mapArguments(this::visit);
        }
        
        @Override
        public Term visitEq(Eq eq) {
            return eq.mapArguments(this::visit);
        }
        
        @Override
        public Term visitApp(App app) {
            return app.mapArguments(this::visit);
        }
        
        private void pushVar(AnnotatedVar av) {
            counter++;
            Mapping m = new Mapping(counter, av.getVar());
            mappingStack.addFirst(m);
        }
        
        private void popVar() {
            counter--;
            mappingStack.removeFirst();
        }
        
        @Override
        public Term visitExists(Exists exists) {
            List<AnnotatedVar> newVars = new ArrayList<>();
            for(AnnotatedVar av : exists.getVars()) {
                pushVar(av);
                newVars.add(Term.mkVar("_" + Integer.toString(counter)).of(av.getType()));
            }
            
            Term body = visit(exists.getBody());
            
            // Return state back to way it was
            for(AnnotatedVar av : exists.getVars()) {
                popVar();
            }
            
            return Term.mkExists(newVars, body);
        }
        
        @Override
        public Term visitForall(Forall forall) {
            List<AnnotatedVar> newVars = new ArrayList<>();
            for(AnnotatedVar av : forall.getVars()) {
                pushVar(av);
                newVars.add(Term.mkVar("_" + Integer.toString(counter)).of(av.getType()));
            }
            
            Term body = visit(forall.getBody());
            
            // Return state back to way it was
            for(AnnotatedVar av : forall.getVars()) {
                popVar();
            }
            
            return Term.mkForall(newVars, body);
        }
    }
    
}

File: core/src/main/scala/fortress/operations/DeBruijnConverter.scala
package fortress.operations

import fortress.msfol._

case class DeBruijnMapping(index: Int, variable: Var)

/** Converts the variables in a term into a standardized DeBruijn Form.
  * We use this to check that substitution works correctly, but don't actually
  * use DeBruijn form within Fortress. */
class DeBruijnConverter {
    private var counter: Int = 0;
    private var mappingStack = List.empty[DeBruijnMapping]
    private val visitor = new DeBruijnVisitor
    
    def convert(term: Term) = visitor.visit(term)
    
    class DeBruijnVisitor extends TermVisitor[Term] {
        
        override def visitTop(): Term = Top
        
        override def visitBottom(): Term = Bottom
        
        override def visitVar(variable: Var): Term = {
            for(m <- mappingStack) {
                if(variable == m.variable) {
                    return Var("_" + Integer.toString(m.index))
                }
            }
            variable
        }
        
        override def visitNot(term: Not): Term = term.mapBody(visit)
        
        override def visitAndList(term: AndList): Term = term.mapArguments(visit)
        
        override def visitOrList(term: OrList): Term = term.mapArguments(visit)
        
        override def visitDistinct(term: Distinct): Term = term.mapArguments(visit)
        
        override def visitImplication(term: Implication): Term = term.mapArguments(visit)
        
        override def visitIff(term: Iff): Term = term.mapArguments(visit)
        
        override def visitEq(term: Eq): Term = term.mapArguments(visit)
        
        override def visitApp(term: App): Term = term.mapArguments(visit)
        
        override def visitBuiltinApp(term: BuiltinApp) = term.mapArguments(visit)

        override def visitClosure(term: Closure) = term.mapArguments(visit)

        override def visitReflexiveClosure(term: ReflexiveClosure) = term.mapArguments(visit)

        private def pushVar(av: AnnotatedVar): Unit = {
            counter += 1
            val m = DeBruijnMapping(counter, av.variable)
            mappingStack = m :: mappingStack
        }
        
        private def popVar(): Unit = {
            counter -= 1
            mappingStack = mappingStack.tail
        }
        
        override def visitExists(term: Exists): Term = {
            val newVars = term.vars.map(av => {
                pushVar(av)
                Var("_" + Integer.toString(counter)) of av.sort
            })
            
            val body = visit(term.body)
            
            // Return state back to way it was
            term.vars.foreach(av => popVar())
            
            Exists(newVars, body)
        }
        
        override def visitForall(term: Forall): Term = {
            val newVars = term.vars.map(av => {
                pushVar(av)
                Var("_" + Integer.toString(counter)) of av.sort
            })
            
            val body = visit(term.body)
            
            // Return state back to way it was
            term.vars.foreach(av => popVar())
            
            Forall(newVars, body)
        }
        
        override def visitDomainElement(d: DomainElement): Term = ???
        
        override def visitIntegerLiteral(literal: IntegerLiteral): Term = ???
        
        override def visitBitVectorLiteral(literal: BitVectorLiteral): Term = ???
        
        override def visitEnumValue(e: EnumValue): Term = ???
        
        override def visitIfThenElse(ite: IfThenElse): Term = ???
        
    }
    
}
